# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/generic/HyperTextAccessible.h
# Commit: 95a2cae69fb9
# Full Hash: 95a2cae69fb9d3763a8a8aba8beba2a78fdd570b
# Author: James Teh <jteh@mozilla.com>
# Date: 2022-05-05 16:11:56
# Regressor Bug: 1766794
# File Overlap Count: 2
# Description:
#   Bug 1766794: Lazily cache HyperText offsets for RemoteAccessibles. r=eeejay
#   
#   We already had a similar cache in local HyperTextAccessible.
#   This improves performance significantly when walking the text attributes of a container with a large number of text leaf children, such as is encountered when using view source.
#   This patch unifies that cache across local and remote.
# ==============================================================================

diff -r 3f8a611b454e -r 95a2cae69fb9 accessible/generic/HyperTextAccessible.h
--- a/accessible/generic/HyperTextAccessible.h	Thu May 05 10:56:32 2022 +0000
+++ b/accessible/generic/HyperTextAccessible.h	Thu May 05 11:14:12 2022 +0000
@@ -159,12 +159,10 @@
 
   virtual already_AddRefed<AccAttributes> DefaultTextAttributes() override;
 
+  virtual void InvalidateCachedHyperTextOffsets() override { mOffsets.Clear(); }
+
   // HyperTextAccessibleBase provides an overload which takes an Accessible.
   using HyperTextAccessibleBase::GetChildOffset;
-  virtual int32_t GetChildOffset(uint32_t aChildIndex,
-                                 bool aInvalidateAfter = false) const override;
-
-  virtual int32_t GetChildIndexAtOffset(uint32_t aOffset) const override;
 
   virtual LocalAccessible* GetChildAtOffset(uint32_t aOffset) const override {
     return LocalChildAt(GetChildIndexAtOffset(aOffset));
@@ -433,11 +431,18 @@
   // HyperTextAccessibleBase
   virtual const Accessible* Acc() const override { return this; }
 
+  virtual const nsTArray<int32_t>& GetCachedHyperTextOffsets() const override {
+    if (mOffsets.IsEmpty()) {
+      BuildCachedHyperTextOffsets(mOffsets);
+    }
+    return mOffsets;
+  }
+
  private:
   /**
    * End text offsets array.
    */
-  mutable nsTArray<uint32_t> mOffsets;
+  mutable nsTArray<int32_t> mOffsets;
 };
 
 ////////////////////////////////////////////////////////////////////////////////